src/Symfony/Component/Security/CHANGELOG.md

```diff
@@ -19,9 +19,7 @@ CHANGELOG
  * Dispatch `AuthenticationFailureEvent` on `security.authentication.failure`
  * Dispatch `InteractiveLoginEvent` on `security.interactive_login`
  * Dispatch `SwitchUserEvent` on `security.switch_user`
- * Added `Argon2idPasswordEncoder`
- * Deprecated using `Argon2iPasswordEncoder` while only the `argon2id` algorithm
-   is supported, use `Argon2idPasswordEncoder` instead
+ * deprecated `Argon2iPasswordEncoder`, use `SodiumPasswordEncoder` instead
```

Comment:
should we actually deprecate it ? It can use the native API of PHP 7.2+ (and Argon2idPasswordEncoder could as well, giving control over the variant). SodiumPasswordEncoder is only based on the sodium extension.

IMO, we should only deprecated the sodium-based fallback in the Argon encoders, not deprecate the encoders themselves.

Reply:
Having encoders bound to one algo is a bad practice. The documented best practice is to have a progressive encoder that knows how to verify old passwords while using the best available algo to encode new ones. So yes: this should be deprecated with this reasnoning.
Of course, there is one missing step: rehashing. See https://github.com/symfony/symfony/issues/31139

Reply:
> The documented best practice is to have a progressive encoder that knows how to verify old passwords while using the best available algo to encode new ones.

The chain / delegating encoder should be composed of encoders using specific algorithms. At least, that's the case in Spring Security (see links above).

Reply:
> The chain / delegating encoder should be composed of encoders using specific algorithms

PHP is a glue language while Java implements everything itself. The situation in PHP is that `password_hash()`/`Sodium\crypto_pwhash_str` already embed the chain and there is zero need to expose the individual algos as encoders.

I agree a chain encoder would be nice to provide a migration path from Pbkdf2/MessageDigest/Plaintext to Native/Sodium thought. But we need #31139 first.

Reply:
> The situation in PHP is that password_hash()/Sodium\crypto_pwhash_str already embed the chain and there is zero need to expose the individual algos as encoders.

Having a default (like with `password_hash`) / fallback (like with `sodium_crypto_pwhash_str`) mechanism does not create a "chain" (to be understood as for the purpose of rehashing). Anyway, I don't understand that argument of PHP vs Java, when it's been pointed out that it's indeed possible to provide encoders for specific algorithms (with extra effort in the case of `sodium_crypto_pwhash`).